### PR TITLE
Add validated flow helper proof of concept

### PR DESCRIPTION
**Pull request by abishekgiri**
## What

- add a trace-backed helper for validating real flow paths from in-memory spans
- add an end-to-end proof of concept for:
  - HTTP trigger
  - `iii::durable::publish`
  - queue trigger
  - `state::set`
  - state trigger
- add a second end-to-end proof of concept for:
  - HTTP trigger
  - `stream::set`
  - stream trigger
- generate renderable flow graph assets only after those runtime paths are actually observed
- serialize access to shared in-memory tracing storage so validated-flow tests remain stable as the helper expands

## Why

- this follows the direction in `#1468` to validate flows from real runtime behavior instead of relying on hand-maintained metadata
- it proves the helper can derive assets from more than one real connection type while staying within the dev/testing-tool boundary
- it keeps the scope intentionally narrow so the model can be reviewed before broader engine or console work

## Notes

- this is still a proof of concept, not a full flows system
- scope is limited to two concrete paths using built-in workers:
  - HTTP -> queue -> state
  - HTTP -> stream trigger
- no console integration or public engine flow API is included yet
- validation used:
  - `cargo fmt --all`
  - `cargo test -p iii --test validated_flow_asset_e2e -- --nocapture`

---

- [x] I am licensing the entirety of this PR under Apache 2 and have all necessary rights to the code I am contributing.

### CODE REVIEW

File: engine/tests/common/flow_helpers.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+            .with(init_otel(&config))
+            .try_init()
+            .expect("flow test tracing subscriber should initialize");
+    });
+
+    if let Some(storage) = get_span_storage() {
+        storage.clear();
+    } else {
+        let _ = InMemorySpanExporter::new(2048, "iii-flow-test".to_string());
+    }
```
[the remark this answers is not in the document]
Comment (abishekgiri) [reply engine/tests/common/flow_helpers.rs:377]: Addressed in the latest push. The helper now fails fast when span storage is missing instead of creating detached storage, so tracing setup issues surface immediately rather than timing out while polling an empty store.

File: engine/tests/validated_flow_asset_e2e.rs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+async fn wait_for_route(client: &reqwest::Client, url: &str) {
+    let deadline = std::time::Instant::now() + Duration::from_secs(3);
+    while std::time::Instant::now() < deadline {
+        if client.get(url).send().await.is_ok() {
+            return;
+        }
+        sleep(Duration::from_millis(10)).await;
+    }
+}
```
[the remark this answers is not in the document]
Comment (abishekgiri) [reply engine/tests/validated_flow_asset_e2e.rs:63]: Addressed in the latest push. wait_for_route now fails at the helper with a clear timeout message, so startup and readiness issues show up at the right point in the test.

File: engine/tests/validated_flow_asset_e2e.rs [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+
+use common::flow_helpers::{HttpQueueStatePathSpec, ensure_flow_test_tracing};
+use common::queue_helpers::builtin_queue_config;
+
+fn reserve_local_port() -> u16 {
+    let listener = StdTcpListener::bind("127.0.0.1:0").expect("bind ephemeral port");
+    let port = listener.local_addr().expect("listener addr").port();
+    drop(listener);
+    port
+}
```
[the remark this answers is not in the document]
Comment (abishekgiri) [reply engine/tests/validated_flow_asset_e2e.rs:30]: Addressed the two smaller follow-ups in the latest push and left this one out for now.

I agree the port reservation pattern is not ideal, but fixing it properly here would likely mean widening the POC into HTTP worker binding and test-harness changes so the test can discover the actual bound port after binding to port 0. I would prefer to keep this PR focused on the validated flow helper proof of concept and handle the port-binding cleanup separately if we want to improve the shared HTTP test harness.

**Comment by abishekgiri** [comment]
Following the plan from #1468, the next step after this POC is to keep the scope narrow and extend the validated-helper pattern to a small number of additional connection types while continuing to derive flow assets from real runtime behavior.

I do not plan to broaden this into console integration or a public engine flows API yet. The goal is to prove the helper shape across a few real execution paths first, then build outward from there once the validated model feels solid.

**Comment by anthonyiscoding** [comment]
@abishekgiri Thanks for opening this PR! Can you address the one remaining comment from coderabbit?

@sergiofilhowz Take a look at this one and let us know what you think about it. I know we're all onboard for flows but I haven't reviewed this code myself.

**Comment by abishekgiri** [comment]
@anthonyiscoding  Addressed the remaining CodeRabbit comment in the latest push.

The flow tracing helper now serializes access to the shared in-memory span storage so future validated-flow tests do not race on `storage.clear()` or span reads as the helper expands. Focused validation still passes with:
- `cargo fmt --all`
- `cargo test -p iii --test validated_flow_asset_e2e -- --nocapture`

File: engine/tests/common/flow_helpers.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            .try_init()
+            .expect("flow test tracing subscriber should initialize");
+    });
+
+    if let Some(storage) = get_span_storage() {
+        storage.clear();
+    } else {
+        panic!("missing span storage: tracing subscriber not installed for flow tests");
+    }
+}
```
[the remark this answers is not in the document]
Comment (abishekgiri) [reply engine/tests/common/flow_helpers.rs]: Addressed in the latest push. The flow tracing helper now acquires a shared async guard before clearing the process-wide span storage, and the test holds that guard for the duration of the run. That keeps validated-flow tests from interfering with each other as this helper expands to additional path types.

**Comment by abishekgiri** [comment]
Expanded the validated-flow proof of concept with the next `#1468` slice in the latest push.

What changed:
- kept the original validated `HTTP -> durable publish -> queue trigger -> state::set -> state trigger` path
- added a second validated `HTTP -> stream::set -> stream trigger` path
- extended the helper so both paths generate deterministic graph assets from real runtime traces and observed events
- serialized access to the shared in-memory span storage so future validated-flow tests do not interfere with each other

Validated with:
- `cargo fmt --all`
- `cargo test -p iii --test validated_flow_asset_e2e -- --nocapture`

**Comment by rohitg00** [comment]
@abishekgiri hey, how's it going? Is this PR up for review for @sergiofilhowz and @andersonleal

**Comment by abishekgiri** [comment]
@rohitg00 Hey, going well, thanks for checking in.

Yes, this is ready for review from the code/design side. This PR came out of the direction discussed in `#1468`, where we moved away from hand-maintained flow metadata as the source of truth and toward validating flow paths from actual runtime behavior.

This first PR is intentionally a proof of concept. It adds a trace-backed test helper that waits for real in-memory spans from an executed iii path, validates that the expected runtime path actually happened, and only then generates a renderable flow graph asset.

Current scope in this PR:
- `HTTP -> iii::durable::publish -> queue trigger -> state::set -> state trigger`
- `HTTP -> stream::set -> stream trigger`
- generated graph asset with nodes and edges from the validated runtime path
- no public engine API changes
- no console integration yet
- no new primitive or workflow DSL

The follow-up work is already split out separately: `#1572` adds the pub/sub path and the reusable declaration model for Phase 1.

Most code checks are passing on this PR. I noticed the license-agreement check is still red, so I can take a look at that separately, but from the implementation side this is ready for review by @sergiofilhowz and @andersonleal.

**Comment by anthonyiscoding** [comment]
Closing for same reasons as #1468. We'll have to get Flows done differently since iii is a fundamentally different system than Motia.
